File: libs/ml-kernel/src/models/experiment.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::models::run::MetricValue;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ModelingObjectiveSpec {
    #[serde(default = "default_objective_status")]
    pub status: String,
    #[serde(default)]
    pub deployment_target: String,
    #[serde(default)]
    pub stakeholders: Vec<String>,
    #[serde(default)]
    pub success_criteria: Vec<String>,
    #[serde(default)]
    pub linked_dataset_ids: Vec<Uuid>,
    #[serde(default)]
    pub linked_model_ids: Vec<Uuid>,
    #[serde(default)]
    pub documentation_uri: String,
    #[serde(default)]
    pub collaboration_notes: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Experiment {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub objective: String,
    pub objective_spec: ModelingObjectiveSpec,
    pub task_type: String,
    pub primary_metric: String,
    pub status: String,
    pub tags: Vec<String>,
    pub run_count: i64,
    pub best_metric: Option<MetricValue>,
    pub owner_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListExperimentsResponse {
    pub data: Vec<Experiment>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateExperimentRequest {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub objective: String,
    #[serde(default = "default_task_type")]
    pub task_type: String,
    #[serde(default = "default_primary_metric")]
    pub primary_metric: String,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub objective_spec: ModelingObjectiveSpec,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UpdateExperimentRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub objective: Option<String>,
    pub task_type: Option<String>,
    pub primary_metric: Option<String>,
    pub status: Option<String>,
    pub tags: Option<Vec<String>>,
    pub objective_spec: Option<ModelingObjectiveSpec>,
}

fn default_task_type() -> String {
    "classification".to_string()
}

fn default_primary_metric() -> String {
    "accuracy".to_string()
}

fn default_objective_status() -> String {
    "draft".to_string()
}

File: libs/ml-kernel-go/models/experiment.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// ModelingObjectiveSpec captures what the experiment is trying to do.
// Defaults: status="draft".
type ModelingObjectiveSpec struct {
	Status              string      `json:"status"`
	DeploymentTarget    string      `json:"deployment_target,omitempty"`
	Stakeholders        []string    `json:"stakeholders,omitempty"`
	SuccessCriteria     []string    `json:"success_criteria,omitempty"`
	LinkedDatasetIDs    []uuid.UUID `json:"linked_dataset_ids,omitempty"`
	LinkedModelIDs      []uuid.UUID `json:"linked_model_ids,omitempty"`
	DocumentationURI    string      `json:"documentation_uri,omitempty"`
	CollaborationNotes  []string    `json:"collaboration_notes,omitempty"`
}

// Experiment is the catalog row for an ML experiment.
type Experiment struct {
	ID            uuid.UUID             `json:"id"`
	Name          string                `json:"name"`
	Description   string                `json:"description"`
	Objective     string                `json:"objective"`
	ObjectiveSpec ModelingObjectiveSpec `json:"objective_spec"`
	TaskType      string                `json:"task_type"`
	PrimaryMetric string                `json:"primary_metric"`
	Status        string                `json:"status"`
	Tags          []string              `json:"tags"`
	RunCount      int64                 `json:"run_count"`
	BestMetric    *MetricValue          `json:"best_metric"`
	OwnerID       *uuid.UUID            `json:"owner_id"`
	CreatedAt     time.Time             `json:"created_at"`
	UpdatedAt     time.Time             `json:"updated_at"`
}

type ListExperimentsResponse struct {
	Data []Experiment `json:"data"`
}

// CreateExperimentRequest defaults: task_type="classification",
// primary_metric="accuracy".
type CreateExperimentRequest struct {
	Name          string                 `json:"name"`
	Description   string                 `json:"description,omitempty"`
	Objective     string                 `json:"objective,omitempty"`
	TaskType      string                 `json:"task_type,omitempty"`
	PrimaryMetric string                 `json:"primary_metric,omitempty"`
	Tags          []string               `json:"tags,omitempty"`
	ObjectiveSpec *ModelingObjectiveSpec `json:"objective_spec,omitempty"`
}

type UpdateExperimentRequest struct {
	Name          *string                `json:"name,omitempty"`
	Description   *string                `json:"description,omitempty"`
	Objective     *string                `json:"objective,omitempty"`
	TaskType      *string                `json:"task_type,omitempty"`
	PrimaryMetric *string                `json:"primary_metric,omitempty"`
	Status        *string                `json:"status,omitempty"`
	Tags          *[]string              `json:"tags,omitempty"`
	ObjectiveSpec *ModelingObjectiveSpec `json:"objective_spec,omitempty"`
}

const (
	DefaultExperimentTaskType      = "classification"
	DefaultExperimentPrimaryMetric = "accuracy"
	DefaultObjectiveStatus         = "draft"
)
